src/tasks.ts

```diff
@@ -96,59 +114,120 @@ export async function main(): Promise<void> {
 
-    debug('Uploading the tree to IPFS');
-    const cid = await uploadTree(tree);
-    debug('Tree uploaded, CID is', cid.toString());
-
     debug('Sending the report');
-    await Shared.ORACLE.connect(Shared.SIGNER).submitReport(
-        targetEpoch,
-        tree.root,
-        distributed,
-        cid.toString(),
+    await Shared.HASHCONSENSUS.connect(Shared.SIGNER).submitReport(
+        refSlot,
+        reportHash,
+        Shared.CONSENSUS_VERSION,
     );
-
     debug('Report sent, storing the artifact');
-    saveArtifact(from, to, tree);
+
+    const { currentFrameConsensusReport } =
+        await Shared.HASHCONSENSUS.getConsensusStateForMember(
+            Shared.SIGNER.address,
+        );
+
+    if (currentFrameConsensusReport == ZeroHash) {
+        debug('No consensus reached');
+        return;
+    }
+
+    if (currentFrameConsensusReport != reportHash) {
+        debug('Oracle`s hash differs from consensus report hash. Exiting');
+        return;
+    }
+
+    // TODO: check for fast lane
+    const treeRoot = await Shared.ORACLE.treeRoot();
+    if (treeRoot == tree.root) {
+        debug('Report already settled');
+        return;
+    }
+
+    debug('Settling the report');
+    debug('Simulating sending report data');
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReportData.staticCall(
+        report,
+        Shared.CONTRACT_VERSION,
+    );
+    debug('Simulation successfull!');
+
+    debug('Sending report data');
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReportData(
+        report,
+        Shared.CONTRACT_VERSION,
+    );
+    debug('Report data sent!');
+    debug('Done');
 }
 
-async function nextReportEpoch(blockTag: BlockTag) {
-    const r = await Shared.ORACLE.nextReportEpoch({
+async function nextReportRefSlot(blockTag: BlockTag): Promise<Slot> {
+    const { refSlot } = await Shared.HASHCONSENSUS.getCurrentFrame({
         blockTag,
     });
 
-    if (!isUint64(r)) {
-        throw new Error('Next report epoch is too big');
+    if (!isUint64(refSlot)) {
+        throw new Error('Next report refSlot is too big');
     }
 
-    return Number(r) as Epoch;
+    return Number(refSlot) as Slot;
 }
 
-async function reportFrame(blockTag: BlockTag) {
-    const r = await Shared.ORACLE.reportFrame({
+async function prevReportRefSlot(blockTag: BlockTag): Promise<Slot> {
+    const refSlot = await Shared.ORACLE.getLastProcessingRefSlot({
         blockTag,
     });
 
-    for (const n of r) {
-        if (!isUint64(n)) {
-            throw new Error(`Report frame slot ${n.toString()} is too big`);
-        }
+    if (!isUint64(refSlot)) {
+        throw new Error('Prev report refSlot is too big');
     }
 
-    return R.map(R.pipe(Number, slotToEpoch))(r) as [Epoch, Epoch];
+    return Number(refSlot) as Slot;
 }
 
-async function loadValidators(epoch: Epoch) {
-    const vals = await getModuleValidators(epochLastSlot(epoch));
+async function loadValidators() {
```

Comment:
removing epoch arg can cause debugging problems, when we need to get a particular response on some epoch previously to figure out what was wrong

Reply:
Currently, it's 'finalized' state being queried, because it will result in less load to the node. I'm not sure one will require to debug actual response from a node. I think for MVP it's fine, and we can reconsider the method to return an epoch number later.